kore/src/Kore/Domain/Builtin.hs

```diff
@@ -205,110 +143,350 @@ instance Unparse child => Unparse (InternalList child) where
         InternalList { builtinListElement } = builtinList
         InternalList { builtinListConcat } = builtinList
 
--- * Builtin Set
+-- * Builtin AC (associative-commutative) generic stuff
 
-{- | Optimized set representation, with elements separated from
-other set terms, and with concrete elements separated from non-concrete ones.
+{- | Internal representation for associative-commutative domain values.
 -}
-data NormalizedSet key child = NormalizedSet
-    { elementsWithVariables :: [child]
-    -- ^ Non-concrete elements of the set. These would be of sort @Int@s for a
-    -- set with @Int@ elements.
-    , concreteElements :: Set key
-    -- ^ Concrete elements of the set. These would be of sort @Int@s for a set
-    -- with @Int@ elements.
-    , sets :: [child]
-    -- ^ Unoptimized (non-element) parts of the set.
+data NormalizedAc key valueWrapper child = NormalizedAc
+    { elementsWithVariables :: [(child, valueWrapper child)]
+    -- ^ Non-concrete elements of the structure.
+    -- These would be of sorts @(Int, String)@ for a map from @Int@ to @String@.
+    , concreteElements :: Map key (valueWrapper child)
+    -- ^ Concrete elements of the structure.
+    -- These would be of sorts @(Int, String)@ for a map from @Int@ to @String@.
+    , opaque :: [child]
+    -- ^ Unoptimized (i.e. non-element) parts of the structure.
     }
-    deriving (Eq, Foldable, Functor, Traversable, GHC.Generic, Ord, Show)
+    deriving (Eq, GHC.Generic, Ord, Show)
 
-instance (Hashable key, Hashable child) => Hashable (NormalizedSet key child)
+instance Functor valueWrapper => Functor (NormalizedAc key valueWrapper) where
```

Comment:
I assume `Functor` and `Foldable` and such are not derivable? I guess that is because of `elementsWithVariables`? It would probably be worthwhile to tweak the data type definition so that it is derivable.